### PR TITLE
LPS-168241 - The user cannot update an ObjectDefinition that has an AccountRestricted

### PR DESCRIPTION
**Pull request by TamyrisTorres**
Related issue: https://issues.liferay.com/browse/LPS-168241

### CODE REVIEW

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectDefinitionLocalServiceImpl.java [hunk 1]
```diff
@@ -1307,7 +1307,8 @@ private void _validateAccountEntryRestrictedObjectFieldId(
 		}
 
 		if (objectDefinition.isApproved() &&
-			objectDefinition.isAccountEntryRestricted()) {
+			(objectDefinition.isAccountEntryRestricted() !=
+				accountEntryRestricted)) {
```
Comment (carolmariaabb) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectDefinitionLocalServiceImpl.java:1311]: Hi @TamyrisTorres I suspect, by reading the error message below `Account entry restriction cannot be disabled when the object definition is published`, that the right condition here is:

```
objectDefinition.isApproved() && 
      objectDefinition.isAccountEntryRestricted() && 
            !accountEntryRestricted
``` 

If `objectDefinition.isAccountEntryRestricted()` is `true` then it cannot be changed to `false`.

Could you please check?

**Review by carolmariaabb — CHANGES_REQUESTED** [review]

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectDefinitionLocalServiceImpl.java (hunk 1 above)
Comment (TamyrisTorres) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectDefinitionLocalServiceImpl.java:1311]: @carolmariaabb I made the change in the message as the customer's need is that we can update the accountEntryRestricted when it has not been published.

Comment (carolmariaabb) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectDefinitionLocalServiceImpl.java:1311]: > as the customer's need is that we can update the accountEntryRestricted when it has not been published.

@TamyrisTorres did you confirm with the product team if this is the expected behavior?

**Comment by anabuchmann** [comment]
Just started reviewing :)

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectDefinitionLocalServiceImpl.java (hunk 1 above)
Comment (TamyrisTorres) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectDefinitionLocalServiceImpl.java]: Hi @carolmariaabb. Yes, I talked about it with Felipe Marinho and we improved the code. Please review the code one more time. Thanks.

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectDefinitionLocalServiceImpl.java [hunk 2]
```diff
@@ -1307,11 +1307,14 @@ private void _validateAccountEntryRestrictedObjectFieldId(
...
+			((accountEntryRestricted !=
+				objectDefinition.isAccountEntryRestricted()) ||
+			 (accountEntryRestrictedObjectFieldId !=
+				 objectDefinition.getAccountEntryRestrictedObjectFieldId()))) {
 
 			throw new ObjectDefinitionAccountEntryRestrictedException(
-				"Account entry restriction cannot be disabled when the " +
-					"object definition is published");
+				"Account entry restriction cannot be altered when the object " +
+					"definition is published");
```
Comment (carolmariaabb) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectDefinitionLocalServiceImpl.java:1317]: Thanks, @TamyrisTorres. In this case, you must also change the UI, by disabling the `accountEntryRestricted` and `accountEntryRestrictedObjectFieldId` fields [here](https://github.com/liferay/liferay-portal/blob/773a927c59f18efadd6c3c137bb61b3e8abb76d8/modules/apps/object/object-web/src/main/resources/META-INF/resources/object_definitions/object_definition/details.jsp#L205-L212).

Right now the, the `accountEntryRestricted` and `accountEntryRestrictedObjectFieldId` fields are disabled if:
 ```
 objectDefinition.isAccountEntryRestricted() && objectDefinition.isApproved()
 ```
 
 But if they must be disabled whenever the object is published, we must remove the  `objectDefinition.isAccountEntryRestricted()` from the condition.

Comment (carolmariaabb) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectDefinitionLocalServiceImpl.java:1317]: In addition, please change from `altered` to `changed`

**Review by carolmariaabb — CHANGES_REQUESTED** [review]

**Comment by TamyrisTorres** [comment]
I checked the AccountRestricted behavior in the UI and understood your observation. I've changed the validation logic to behave the same as the UI. Please do one more review.

**Comment by naraandrade** [comment]
ci:test:sf

**Comment by naraandrade** [comment]
ci:test:relevant

**Review by carolmariaabb — APPROVED** [review]
Hi @TamyrisTorres,

I'm going to approve this PR, but just to be clear, the final behavior will be:

> Account entry restriction cannot be disabled when the object definition is published

Correct?

**Comment by carolmariaabb** [comment]
ci:reevaluate:1257401_5574

**Comment by TamyrisTorres** [comment]
> 

Yup. is correct.

**Comment by carolmariaabb** [comment]
Manually forwarded https://github.com/brianchandotcom/liferay-portal/pull/127049

**Pull request closed without merging**